capsules/src/ble_advertising_driver.rs

```diff
@@ -338,22 +338,25 @@ where
     // likely be chosen.
     fn reset_active_alarm(&self) {
         let now = self.alarm.now();
-        let mut next_alarm = u32::max_value();
+        let mut next_ref = u32::max_value();
+        let mut next_dt = u32::max_value();
         let mut next_dist = u32::max_value();
         for app in self.app.iter() {
             app.enter(|app, _| match app.alarm_data.expiration {
-                Expiration::Abs(exp) => {
-                    let t_dist = exp.wrapping_sub(now);
+                Expiration::Enabled(reference, dt) => {
+                    let exp = reference.wrapping_add(dt);
+                    let t_dist = exp.wrapping_sub(now.into_u32());
                     if next_dist > t_dist {
-                        next_alarm = exp;
+                        next_ref = reference;
+                        next_dt = dt;
                         next_dist = t_dist;
                     }
                 }
                 Expiration::Disabled => {}
             });
         }
-        if next_alarm != u32::max_value() {
-            self.alarm.set_alarm(next_alarm);
+        if next_ref != u32::max_value() {
```

Comment:
Why not use an option here?

Reply:
Because the previous implementation didn't and I was trying to minimize the changes. It would make a lot of sense (although use an additional word of memory).

Reply:
You have almost 3000 lines of additions in a single PR, I don't think converting this to an option would be too much change :P

Reply:
Sure, happy to change it, I just wanted to scope this as "updating everything to the new abstractions" not "rewriting code to be more idiomatic." Given that it is 3,000 lines, asking people to also reason about the correctness of rewrites for different programming style seemed to me that it could be the straw that broke the camel's back.

Although utltmately at this point the hope is to get a detailed review of the abstractions and for people to see how those abstractions manifest in code. It's intended to be a final design review more than a specific code review.